File: server/src/quotas/minimax.ts
// server/src/quotas/minimax.ts — MiniMax 编码订阅额度（移植 CodexBar MiniMaxUsageFetcher 的 remains 口径）：
// 凭据 = env MINIMAX_CODING_API_KEY / MINIMAX_API_KEY；
// GET {apiBase}/v1/token_plan/remains → 回退 /v1/api/openplatform/coding_plan/remains，
// 国际（minimax.io）失败换中国（minimaxi.com）；*_usage_count 字段是剩余量而非已用量。
import {
  finalizeAccount, missingAccount, quotaFetchJson, readHttpStatus, errText,
  quotaNum, quotaStr, percentOf, toEpochMs,
  type FetchLike, type QuotaAccount, type QuotaWindow,
} from './types.js'

const LABEL = 'MiniMax'

type Region = { apiBase: string }

const REGION_INTL: Region = { apiBase: 'https://api.minimax.io' }
const REGION_CN: Region = { apiBase: 'https://api.minimaxi.com' }

export function readMiniMaxAuth(env: NodeJS.ProcessEnv): { token: string; regions: Region[] } | null {
  const token = env['MINIMAX_CODING_API_KEY']?.trim() || env['MINIMAX_API_KEY']?.trim() || ''
  if (!token) return null
  const primary = env['MINIMAX_REGION']?.trim().toLowerCase() === 'cn' ? REGION_CN : REGION_INTL
  const other = primary === REGION_CN ? REGION_INTL : REGION_CN
  return { token, regions: [primary, other] }
}

/** 单条 model_remains → 窗口；intervals/weekly 的 usage_count 字段是剩余量 */
function windowFromItem(
  key: string, label: string, item: Record<string, unknown>,
  totalKey: string, remainingKey: string, remainingPercentKey: string, endKey: string,
): QuotaWindow | null {
  const total = quotaNum(item[totalKey])
  const remainingRaw = quotaNum(item[remainingKey])
  const remainingPercent = quotaNum(item[remainingPercentKey])
  const used = total !== null && remainingRaw !== null
    ? Math.max(0, total - remainingRaw)
    : null
  const percentage = percentOf(used, total)
    ?? (remainingPercent !== null ? Math.min(100, Math.max(0, 100 - remainingPercent)) : null)
  if (percentage === null && total === null) return null
  return {
    key, label,
    total, used, remaining: remainingRaw,
    percentage,
    nextResetAt: toEpochMs(item[endKey]),
  }
}

export async function fetchMiniMaxAccount(
  auth: { token: string; regions: Region[] }, doFetch: FetchLike, now: number,
): Promise<QuotaAccount> {
  const headers: Record<string, string> = {
    authorization: `Bearer ${auth.token}`,
    accept: 'application/json',
    'content-type': 'application/json',
    'mm-api-source': 'wattson-usage',
  }
  let body: unknown = null
  let lastErr: unknown = null
  let httpStatus: number | null = null
  outer: for (const region of auth.regions) {
    for (const path of ['/v1/token_plan/remains', '/v1/api/openplatform/coding_plan/remains']) {
      try {
        body = (await quotaFetchJson(doFetch, `${region.apiBase}${path}`, { headers })).body
        lastErr = null
        break outer
      } catch (err) {
        lastErr = err
        httpStatus = readHttpStatus(err)
      }
    }
  }
  if (lastErr !== null || body === null) {
    return finalizeAccount({
      kind: 'minimax', label: LABEL, provider: null,
      error: errText(lastErr), httpStatus, now,
    })
  }
  try {
    const root = (body ?? {}) as Record<string, any>
    const data = (root['data'] ?? {}) as Record<string, any>
    const baseResp = (data['base_resp'] ?? root['base_resp'] ?? null) as Record<string, any> | null
    const statusCode = quotaNum(baseResp?.['status_code'])
    if (statusCode !== null && statusCode !== 0) {
      const message = quotaStr(baseResp?.['status_message']) ?? `status_code ${statusCode}`
      const err: Error & { quotaHttpStatus?: number } = new Error(message)
      if (statusCode === 1004) err.quotaHttpStatus = 401
      throw err
    }
    const items = Array.isArray(data['model_remains']) ? data['model_remains'] : []
    if (items.length === 0) throw new Error('响应缺少 model_remains 数据')

    // 多模型 lane 取最吃紧的一个窗口口径
    let interval: QuotaWindow | null = null
    let week: QuotaWindow | null = null
    for (const raw of items) {
      if (raw === null || typeof raw !== 'object') continue
      const item = raw as Record<string, unknown>
      const candidateInterval = windowFromItem(
        'fiveHour', '5 小时窗口', item,
        'current_interval_total_count', 'current_interval_usage_count',
        'current_interval_remaining_percent', 'end_time',
      )
      if (candidateInterval && (interval === null || (candidateInterval.percentage ?? 0) > (interval.percentage ?? 0))) {
        interval = candidateInterval
      }
      const candidateWeek = windowFromItem(
        'week', '周额度', item,
        'current_weekly_total_count', 'current_weekly_usage_count',
        'current_weekly_remaining_percent', 'weekly_end_time',
      )
      if (candidateWeek && (week === null || (candidateWeek.percentage ?? 0) > (week.percentage ?? 0))) {
        week = candidateWeek
      }
    }
    const windows: QuotaWindow[] = []
    if (interval) windows.push(interval)
    if (week) windows.push(week)
    return finalizeAccount({
      kind: 'minimax', label: LABEL, provider: null,
      windows, anySuccess: true, now,
    })
  } catch (err) {
    return finalizeAccount({
      kind: 'minimax', label: LABEL, provider: null,
      error: errText(err), httpStatus: readHttpStatus(err), now,
    })
  }
}

export function minimaxAccount(_home: string, env: NodeJS.ProcessEnv, doFetch: FetchLike, now: number): QuotaAccount | Promise<QuotaAccount> {
  const auth = readMiniMaxAuth(env)
  if (!auth) return missingAccount('minimax', LABEL, now)
  return fetchMiniMaxAccount(auth, doFetch, now)
}

File: Sources/WattsonCore/Quotas/MiniMax.swift
// Quotas/MiniMax.swift — MiniMax 编码订阅额度（移植自原生实现 MiniMaxUsageFetcher 的 remains 口径）：
// 凭据 = env MINIMAX_CODING_API_KEY / MINIMAX_API_KEY；
// GET {apiBase}/v1/token_plan/remains → 回退 /v1/api/openplatform/coding_plan/remains，
// 国际（minimax.io）失败换中国（minimaxi.com）；*_usage_count 字段是剩余量而非已用量。
import Foundation

private let LABEL = "MiniMax"

public struct MiniMaxRegion: Sendable, Equatable {
    public let apiBase: String
    public init(apiBase: String) { self.apiBase = apiBase }
}

private let REGION_INTL = MiniMaxRegion(apiBase: "https://api.minimax.io")
private let REGION_CN = MiniMaxRegion(apiBase: "https://api.minimaxi.com")

/// 凭据：env MINIMAX_CODING_API_KEY → MINIMAX_API_KEY 回退；MINIMAX_REGION=cn 决定首选区域
public func readMiniMaxAuth(_ env: [String: String]) -> (token: String, regions: [MiniMaxRegion])? {
    let coding = env["MINIMAX_CODING_API_KEY"]?.trimmingCharacters(in: .whitespaces) ?? ""
    let generic = env["MINIMAX_API_KEY"]?.trimmingCharacters(in: .whitespaces) ?? ""
    // TS `?.trim() || ?.trim() || ''`：空串视为缺省，向后回退
    let token = !coding.isEmpty ? coding : generic
    if token.isEmpty { return nil }
    let primary = env["MINIMAX_REGION"]?.trimmingCharacters(in: .whitespaces).lowercased() == "cn"
        ? REGION_CN : REGION_INTL
    let other = primary == REGION_CN ? REGION_INTL : REGION_CN
    return (token, [primary, other])
}

/// 单条 model_remains → 窗口；intervals/weekly 的 usage_count 字段是剩余量
func minimaxWindowFromItem(key: String, label: String, item: [String: JSON],
                           totalKey: String, remainingKey: String,
                           remainingPercentKey: String, endKey: String) -> QuotaWindow? {
    let total = quotaNum(item[totalKey])
    let remainingRaw = quotaNum(item[remainingKey])
    let remainingPercent = quotaNum(item[remainingPercentKey])
    let used: Double?
    if let t = total, let r = remainingRaw { used = max(0, t - r) } else { used = nil }
    let percentage = percentOf(used, total)
        ?? remainingPercent.map { min(100, max(0, 100 - $0)) }
    if percentage == nil && total == nil { return nil }
    return QuotaWindow(key: key, label: label,
                       total: total, used: used, remaining: remainingRaw,
                       percentage: percentage, nextResetAt: toEpochMs(item[endKey]))
}

public func fetchMiniMaxAccount(_ auth: (token: String, regions: [MiniMaxRegion]), _ doFetch: FetchLike, _ now: Double) async -> QuotaAccount {
    let headers = [
        "authorization": "Bearer \(auth.token)",
        "accept": "application/json",
        "content-type": "application/json",
        "mm-api-source": "wattson-usage",
    ]
    var body: JSON?
    var lastErr: Error?
    var httpStatus: Double?
    // 双层回退：区域 × 两条 path，任一成功即停（外层 break 语义）
    outer: for region in auth.regions {
        for path in ["/v1/token_plan/remains", "/v1/api/openplatform/coding_plan/remains"] {
            do {
                body = try await quotaFetchJson(doFetch, "\(region.apiBase)\(path)", headers: headers).body
                lastErr = nil
                break outer
            } catch {
                lastErr = error
                httpStatus = readHttpStatus(error).map(Double.init)
            }
        }
    }
    if lastErr != nil || body == nil {
        return finalizeAccount(kind: .minimax, label: LABEL,
                               error: lastErr.map(errText), httpStatus: httpStatus, now: now)
    }
    do {
        let root = body ?? .null
        let data = root["data"] ?? .null
        let baseResp = data["base_resp"] ?? root["base_resp"]
        let statusCode = quotaNum(baseResp?["status_code"])
        if let statusCode, statusCode != 0 {
            let message = quotaStr(baseResp?["status_message"])
                ?? "status_code \(statusCode.truncatingRemainder(dividingBy: 1) == 0 ? String(Int(statusCode)) : "\(statusCode)")"
            if statusCode == 1004 {
                throw HTTPStatusError(status: 401, message: message)
            }
            throw HTTPStatusError(status: 0, message: message)
        }
        let items = data["model_remains"]?.arr ?? []
        if items.isEmpty { throw QuotaMessageError(message: "响应缺少 model_remains 数据") }

        // 多模型 lane 取最吃紧的一个窗口口径
        var interval: QuotaWindow?
        var week: QuotaWindow?
        for raw in items {
            guard let obj = raw.obj else { continue }
            if let candidate = minimaxWindowFromItem(
                key: "fiveHour", label: "5 小时窗口", item: obj,
                totalKey: "current_interval_total_count", remainingKey: "current_interval_usage_count",
                remainingPercentKey: "current_interval_remaining_percent", endKey: "end_time") {
                if interval == nil || (candidate.percentage ?? 0) > (interval?.percentage ?? 0) {
                    interval = candidate
                }
            }
            if let candidate = minimaxWindowFromItem(
                key: "week", label: "周额度", item: obj,
                totalKey: "current_weekly_total_count", remainingKey: "current_weekly_usage_count",
                remainingPercentKey: "current_weekly_remaining_percent", endKey: "weekly_end_time") {
                if week == nil || (candidate.percentage ?? 0) > (week?.percentage ?? 0) {
                    week = candidate
                }
            }
        }
        var windows: [QuotaWindow] = []
        if let interval { windows.append(interval) }
        if let week { windows.append(week) }
        return finalizeAccount(kind: .minimax, label: LABEL,
                               windows: windows, anySuccess: true, now: now)
    } catch {
        return finalizeAccount(kind: .minimax, label: LABEL,
                               error: errText(error), httpStatus: readHttpStatus(error).map(Double.init), now: now)
    }
}

public func minimaxAccount(_ home: String, env: [String: String], doFetch: FetchLike, now: Double) async -> QuotaAccount {
    // home 未用：MiniMax 凭据只来自 env，仅为统一签名保留
    guard let auth = readMiniMaxAuth(env) else { return missingAccount(.minimax, LABEL, now) }
    return await fetchMiniMaxAccount(auth, doFetch, now)
}
